### ISSUE TITLE
SIGSEGV using jep from specs2 in sbt - embedded both Python (jep) and R (rpy2)

### ISSUE THREAD

**Issue by SemanticBeeng** [ninia/jep#126]
- OS Platform, Distribution, and Version: ubutntu 16
- Python Distribution and Version: 3.6.5
- Java Distribution and Version: Oracle Jdk 8
- Jep: from v3.7.1 tag

Getting this from application test code.
Using `specs2` to run `jep` which calls a simple python script.

From sbt, first time it runs ok but second time it cashes with segmentation violation.
Any thoughts please @ndjensen ?

It feels like sbt jvm holds the spec class loaded and this, in turn, holds the python binaries ... hanging around.
I use something like 
```
export LD_PRELOAD=$DSENV/lib/libpython"${PY_VER}"m.so
```
and also set `LD_LIBRARY_PATH`.

Thought it might be related to this https://github.com/ninia/jep/issues/22

```
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGSEGV (0xb) at pc=0x00007f40cae294e0, pid=14851, tid=0x00007f404a0dc700
#
# JRE version: Java(TM) SE Runtime Environment (8.0_131-b11) (build 1.8.0_131-b11)
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.131-b11 mixed mode linux-amd64 compressed oops)
# Problematic frame:
# C  [libpython3.6m.so+0x19d4e0]  threadstate_getframe+0x0
#
# Core dump written. Default location: /development/projects/.../foundation/core or core.14851
#
# An error report file with more information is saved as:
# /development/projects/.../hs_err_pid14851.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
# The crash happened outside the Java Virtual Machine in native code.
# See problematic frame for where to report the bug.
```

```
package com.comp.datalayer

import io.comp.datalayer.Code
import io.comp.jep.Scalathon
import jep.Jep
import org.slf4j.LoggerFactory
import org.specs2.execute.{AsResult, Result}
import org.specs2.mutable.Specification
import org.specs2.specification.ForEach

trait callContext extends ForEach[Jep] {


  def foreach[R: AsResult](f: Jep => R): Result = {

    implicit val jep: Jep = new Jep()
    try {
      println(s"spec start run ${jep} on thread ${Thread.currentThread().getName}")
      AsResult(f(jep))
    } finally {
      println(s"spec end end ${jep} on thread ${Thread.currentThread().getName}")
      jep.close()
    }
  }
}


class AnalysisOrchestrationSpec extends Specification with callContext {

  sequential

  protected val logger = LoggerFactory.getLogger("io.comp.orchestration")

  "orchestration spec" should {

    "extract some feature" >> { jep: Jep =>

      runExtractFeature1(jep)
      ok
    }

  }

  private def runExtractFeature1(implicit jep: Jep): Unit = {

    import Scalathon._

    pythonImport(s"${Code.path}/extract_features")
    val feature1: Long = 'extract_feature1 (Seq(100))
    println(feature1)

  }
}
```

```
from java.lang import Integer
from java.util import ArrayList as AL

def extract_feature1(param1):

    print(f">>>>>>> Extracted feature {param1}")

    x = Integer(5)
    y = Integer(51)
    a = AL()

    return 1
```

Reused code from https://github.com/A-OK/Snakes-and-Ladders/blob/master/scala_hosting_python/jepp/a_la_spiewak/Scalathon.scala

**Comment by SemanticBeeng** [comment]
called under finally; see second `println`

**Comment by bsteffensmeier** [comment]
I'm not sure what you mean, `isValidThread` cannot be the cause of the crash since it makes no native calls so it cannot call into libpython where the crash occured.

What is the result of your printlns? Is close() getting called on the same thread as creation. Do you see any of the warnings that are [here](https://github.com/ninia/jep/blob/master/src/jep/Jep.java#L1195) or [here](https://github.com/ninia/jep/blob/master/src/jep/Jep.java#L393)?

If you can include the java frames and native frames from the hs_err_pid file that might be helpful.

**Comment by SemanticBeeng** [comment]
thanks! will follow up in 8 hrs

**Comment by SemanticBeeng** [comment]
Did some research into potentially related issues

https://github.com/ninia/jep/issues/28 
> _"ep3.4.1 on Mac in a Scala environment, with scipy. If I try to initialize multiple jep instances and import scipy, I get an error the second time."_

https://github.com/ninia/jep/issues/28#issuecomment-136917996
> _" This is probably a scipy issue and not a Jep issue. The project I work on uses scipy but not extensively (uses numpy much more),"_

> _"Jep can't really do much about this kind of thing if the CPython code is messing up internal function pointers, either at python sub-interpreter shutdown or the start of another sub-interpreter, that's in the specific extension's code, not Jep"_

In my case, likely `rpy2` does this sort of thing.

https://github.com/ninia/jep/issues/28#issuecomment-137600583
> _"was thinking making independent instances of Jep per thread protects the variables and namespaces of a script but obviously this dis not seem to be the case. The way I got around this problem was by making synchronized calls to jep.invoke"_

https://github.com/ninia/jep/issues/28#issuecomment-150028479
> _"The root of the problem is in the numpy c code where it imports the _mehods module into c and save methods to a static variables"_

Is this on the right direction ?

**Comment by SemanticBeeng** [comment]
Introduced use of shared modules and this has improved stability

```
  val config = new JepConfig().addSharedModules("numpy","rpy2")

  def foreach[R: AsResult](f: Jep => R): Result = {

    println("Initializing Jep ...")
    implicit val jep: Jep = new Jep(config)
```

Not a combination of the above involving `numpy` and `rpy2` + R code execution does not cause SIGSEGV in Intellij.

In Intellij still happens on the second run.

**Comment by bsteffensmeier** [comment]
When you start mixing in numpy and rpy2 it is not surprising that you can create conditions that crash. Jep is built off of python sub-interpreters which are a rarely used feature of cpython so extension writers often write code that is not compatible with sub-interpreters. There is not much we can do about this from the jep, we have open issues with numpy but it is fairly complex to fix them and there may be performance side affects so there is no activity to fix them. Our current workaround is the introduction of shared modules and in the next release of jep(3.8) I have added an option to use jep without sub-interpreters which should also help alleviate this problem but if you get the configuration wrong with native python extensions it can crash.

What was most interesting in the original post was that you managed to crash without using much python at all. I don't know how sbt works, but I think you were on the right track when you mentioned sbt keeping the python binaries around. I suspect that sbt is using the same process for both runs, however it attempted to unload the Jep class and reload it which causes it to reinitialize the python interpreter. So the first question is if you can verify this? Can you print out the process ID or something to verify that both runs are in the same process?

In theory Jep attempts to support this use case, when the first jep class is unloaded the JVM should call JNI_OnUnload, [which calls pyembed_shutdown](https://github.com/ninia/jep/blob/dev_3.7/src/jep/jep.c#L57), which should completely reset the python state. However I have never managed to get the JVM to actually unload in my testing. I suspect that the JVM cannot unload the Jep class because the [TopInterpreter Thread](https://github.com/ninia/jep/blob/dev_3.7/src/jep/Jep.java#L132) is still running. Looking at your hs_err_pid, I suspect that ` 0x00007f6edc01c800 JavaThread "Thread-1" daemon` is actually the first jep TopInterpreter still running and `0x00007f6ed80f3800 JavaThread "Thread-2" daemon` is the second TopInterpreter which crashes when it tries to reinitialize python. Although looking at the  initialization code, [this check](https://github.com/ninia/jep/blob/dev_3.7/src/jep/pyembed.c#L230) should actually prevent a crash during initialization, so I'm still not really sure exactly what is going on.

I suspect this problem may be resolved if you use forked JVMs in your SBT configuration, reading [some documentation](https://www.scala-sbt.org/1.0/docs/Running-Project-Code.html) it looks like the Jep background thread is exactly the type of use case that needs forking. It would still be good if we could fix this within Jep so please post any other information you find, but it would not be trivial if we really do need to handle jep class unloading so forking the JVM may be the most practical option.

**Comment by SemanticBeeng** [comment]
_" managed to crash without using much python "_ - I believe it was from `import pandas2ri`
At some point I managed to avoid crash by simply commenting that out.

Until I started using `rpy2` in shared module could not get rid of SIGSEGV. 
And after I used them I never got a SIGSEGV in Intellij.

It would be great of @lgautier had a look at this since he is master of running running R embedded (so he might help with running Python embedded).

----
side note: At some point I had a script using `rpy2` and failing to convert a `pandas` frame to R just because `pandas` was not loaded as shared module also:

```
r_dataframe = pandas2ri.py2ri(pdf)
```

**Comment by SemanticBeeng** [comment]
> _" current workaround is the introduction of shared modules and in the next release of jep(3.8) I have added an option to use jep without sub-interpreters "_

What is the ETA please?

**Comment by SemanticBeeng** [comment]
> _the right track when you mentioned sbt keeping the python binaries around_
Wip investigating with forking but no joy yet.
Seems there is not way to get PID from java before Java 9.

**Comment by bsteffensmeier** [comment]
> What is the ETA please?

The code to run without subinterpreters is already in the dev_3.8 branch if you want to try it out, see (here)[https://github.com/ninia/jep/blob/dev_3.8/src/main/java/jep/JepConfig.java#L104]. The official release of 3.8 will probably happen sometime around June or July.

**Comment by SemanticBeeng** [comment]
Forking of tests in sbt project works and ~solves~ goes around the issue of SIGSEGV in second run in interactive sbt session !

```
Initializing Jep in PID=21703 ...
Start running spec in Jep=jep.Jep@48c7e05a / thread=specs2-3
...
Finish running spec in Jep=jep.Jep@48c7e05a / thread=specs2-3
...
Initializing Jep in PID=21703 ...

Start running spec in Jep=jep.Jep@29b64861 / thread=specs2-2
...
Finish running spec in Jep=jep.Jep@29b64861 / thread=specs2-2
```

**Comment by SemanticBeeng** [comment]
Found another way to break it

```
  private def runNumpyTest(implicit jep1: Jep): Unit = {

    jep1.close()

    var jep = new Jep()
    jep.eval("import numpy")
    jep.eval("numpy.ndarray([1]).any()")
    jep.close()

    jep = new Jep()
    jep.eval("import numpy")  // <+++++ SIGSEGV here
    jep.eval("numpy.ndarray([1]).any()")
    jep.close()
  }
```

```
# Problematic frame:
# C  [multiarray.cpython-36m-x86_64-linux-gnu.so+0x13ea38]  PyArray_Item_INCREF+0xe8
#
```

Inspired from something Nate did.
Do not really need this to work - hope not! - but might help narrow down the cause.

**Comment by lgautier** [comment]
@bsteffensmeier 

> Jep is built off of python sub-interpreters which are a rarely used feature of cpython so extension writers often write code that is not compatible with sub-interpreters.

My understanding is that [C-extensions are shared across the sub-interpreters](https://docs.python.org/3/c-api/init.html#c.Py_NewInterpreter), and because of this the exact way to develop with sub-interpreters in mind is a non-trivial affair (and an improvement the subject of a PEP under discussion - https://www.python.org/dev/peps/pep-0554/).

**Comment by lgautier** [comment]
@SemanticBeeng 

`import rpy2.rinterface` does not initialize the embedded while `import rpy2.robjects.pandas2ri` does when implicitly importing `rpy2.robjects`. If you want to decompose steps, consider calling `rinterface.initr()` before importing `pandas2ri` (see the [doc](https://rpy2.github.io/doc/v2.9.x/html/rinterface.html#initialization)).

**Comment by lgautier** [comment]
If the C-extension is shared as (I understand) the doc tells it, calling `initr()` only once should work.

You can check it easily, as `rpy2` is maintaining R's initialization status (since it is unfortunately not part of R's C API):

```python
rpy2.rinterface._rinterface.is_initialized()
```

However, calling `initr()` multiple times should not matter, unless initialization options are changed between calls and the code relies on the changes: R will only be initialized the first time `initr()` is called.

**Comment by bsteffensmeier** [comment]
> Do not really need this to work - hope not! - but might help narrow down the cause

If you close an interpreter using numpy and you are not using shared modules then there are many ways to crash or generate errors. This is caused by numpy storing static references to python objects, there is nothing we can do in jep so you don't need to report it here.

**Comment by bsteffensmeier** [comment]
> My understanding is that C-extensions are shared across the sub-interpreters, and because of this the exact way to develop with sub-interpreters in mind is a non-trivial affair (and an improvement the subject of a PEP under discussion - https://www.python.org/dev/peps/pep-0554/).

The complexity depends on what you are trying to do, but I agree for many things it may be non-trivial. Most of my experience has been digging in numpy which has problems because they keep static references in c to python objects. When a python interpreter is closed these objects become invalid and fail if numpy is used again. [PEP-3121](https://www.python.org/dev/peps/pep-3121/) was intended to provide extension writers away around this particular problem but it isn't widely used. Pep-554 is interesting so I'll be curious to see how that changes things in the future.

**Comment by SemanticBeeng** [comment]
> _"close an interpreter using numpy and you are not using shared modules"_

Understood and am using shared modules.

At this time this code does not cause SIGSEGV. 

```
  protected def runNumpyTest(implicit jep: Jep): Int = {

    printContext

    jep.eval("import numpy")
    jep.eval("numpy.ndarray([1]).any()")

    jep.eval("import numpy")
    jep.eval("numpy.ndarray([1]).any()")
    0
  }
```
Will test more closing `jep`  instances in relation to `numpy` (with shared modules).
If it is not too much to ask, please post links to pending `numpy` issues from `jep` perspective.

**Comment by bsteffensmeier** [comment]
[Numpy issues labeled embedded](https://github.com/numpy/numpy/issues?utf8=%E2%9C%93&q=is%3Aissue+is%3Aopen+label%3AEmbedded+)

**Comment by dlong11** [comment]
Just for the record - I experienced the same issue with numpy - https://github.com/numpy/numpy/issues/13051#issuecomment-509713683  - I got a response but appears that it is a low priority.

**Comment by bsteffensmeier** [comment]
I am closing this old issue because it is specific to a very specific environment and there have been many changes to the jep, python, and numpy since we last looked at this. I do not see any changes we need to make in jep in this discussion. If  I am mistaken and we need jep changes please open a new issue.
